Automattic/wp-calypso
# Memberships: Add coupons GUI

Related to 225-gh-automattic/gold

## Proposed Changes

* Adding a coupons GUI for memberships.

## Translators

You should be able to find each of the strings in these screenshots


### Coupons list

- `%s off for 1 month`
- `%s off for 3 months`
- `%s off for 6 months`
- `%s off for 1 year`
- `%s off forever`
- `Manage coupons`
- `Add a new coupon`
- `First-time order only`
- `Limited to specific emails`
- `Limited to specific products`

![Screenshot 2023-12-29 at 12 52 09 PM](https://github.com/Automattic/wp-calypso/assets/3662616/31e5f27b-4ad5-4128-85dd-b42587e5e752)

### Add Coupon

#### Any product
- `Any product`
- `Choose a unique coupon code for the discount. Not case-sensitive.`
- `Percentage`
- `Expiration date (optional)`
- `Usage limit (optional)`
- `Limit the number of times this coupon can be redeemed by a supporter.`
- `Eligible for first-time order only`
- `Once the coupon is used for a subscription, how long does it last?`
- `Limit coupon to specific emails`
- `Separate email addresses with commas. Use an asterisk (*) to match parts of an email. For example, "*@university.edu" would select all "university.edu" addresses.`

![Screenshot 2023-12-29 at 12 52 22 PM](https://github.com/Automattic/wp-calypso/assets/3662616/85abc844-cd86-4245-a0af-9a981f207e02)

#### 1 Product Selected
- `1 product selected`

![Screenshot 2023-12-29 at 12 52 55 PM](https://github.com/Automattic/wp-calypso/assets/3662616/f36aee6d-c97c-414e-960d-977d37462398)

#### Multiple products selected
- `%s products selected.`

![Screenshot 2023-12-29 at 12 53 16 PM](https://github.com/Automattic/wp-calypso/assets/3662616/eb62cc0e-bedf-4724-986d-c5b70b05bdc3)

#### Duration
- `3 Months`
- `6 Months`
- `1 Year`

![Screenshot 2023-12-29 at 1 04 55 PM](https://github.com/Automattic/wp-calypso/assets/3662616/21700445-35fe-4b91-97e1-2c9d38e22505)

#### Errors
NOTE: To get these error message, click or tab into the input field, then tab or click away without entering any information, or by entering purposefully invalid information (negative numbers, end dates before start dates, etc..)

##### Invalid coupon code
- `Please input a coupon code.`

![Screenshot 2023-12-29 at 1 06 16 PM](https://github.com/Automattic/wp-calypso/assets/3662616/04e45696-93a0-4508-9791-1e9271e03315)

##### Invalid type
This error is difficult to trigger, which is why there is no screenshot. However, it will look like the other errors and will be under the "Discount type" selector.
- `Please select a valid type`

##### Invalid amount
- `Please enter a valid amount`

![Screenshot 2023-12-29 at 1 13 52 PM](https://github.com/Automattic/wp-calypso/assets/3662616/c82475b7-1053-4070-b288-ff1719fd3b6c)

##### Invalid dates
- `Please enter a valid start date`
- `Please enter a valid end date`

![Screenshot 2023-12-29 at 1 14 59 PM](https://github.com/Automattic/wp-calypso/assets/3662616/785ce2f8-3722-4235-9645-081c65a1c032)

##### Invalid usage limit
- `Please enter a positive value, or 0 for infinite uses.`

![Screenshot 2023-12-29 at 1 15 42 PM](https://github.com/Automattic/wp-calypso/assets/3662616/df092beb-cf89-4a22-a4a1-c24773c14c29)

##### Invalid emails
- `Please fix these invalid email addresses: %s`

![Screenshot 2023-12-29 at 1 16 06 PM](https://github.com/Automattic/wp-calypso/assets/3662616/0638f938-2912-40a1-a116-4af6ac51f7f9)

### Delete coupon?
- `Delete coupon?`
- `Deleting this coupon ({{strong}}%s{{/strong}}) will prevent coupon holders from applying coupon at checkout.`
![Screenshot 2023-12-29 at 12 53 49 PM](https://github.com/Automattic/wp-calypso/assets/3662616/a2355089-e085-4ba8-9d1a-504eac99ef95)

## Notifications
_NOTE: to get these notifications, you need to save a new coupon or update an existing coupon_

### Added Coupon
- `Added coupon "%s".`

![Screenshot 2023-12-29 at 1 16 34 PM](https://github.com/Automattic/wp-calypso/assets/3662616/d568a8f0-4d8c-49c0-9364-fb8c8d4a3f78)

### Updated Coupon
- `Updated coupon "%s".`

![Screenshot 2023-12-29 at 1 17 37 PM](https://github.com/Automattic/wp-calypso/assets/3662616/3db3dd5e-0fc5-4d64-b30d-8b13998862ff)

## Testing Instructions



* Stick the appropriate blog sticker to make coupons work on your memberships site. Hint: search wpcom for `is_coupons_and_gifts_enabled`. This allows the coupons to be used with the backend, otherwise the routes will not exist.
* Apply D133250-code to your sandbox. Sandbox the public API and subscribe.wordpress.com. Enable store sandbox.
* Navigate to `/earn/payments/` using one of the calypso live links below.
* Click "Add coupon" and add a coupon. Try various options. Click save.
* Edit the coupons by clicking the ellipsis menu and clicking edit. The same goes for deleting them.
* Create more coupons with different configurations and ensure they work as expected.

## Pre-merge Checklist
